src/System.Slices/src/System/Contract.cs

```diff
@@ -31,17 +31,26 @@ public static void Requires(bool condition)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void RequiresInRange(int start, int length)
         {
-            if (!(start >= 0 && start < length))
+            // explicit unchecked block is used in order to avoid OverflowException when casting 
+            // negative int to unsigned int when the code is compiled with /checked option
+            unchecked
             {
-                throw new ArgumentOutOfRangeException();
+                // use unsigned int to reduce range check by one ( >= 0 )
+                if ((uint)start >= (uint)length)
```

Comment:
@whoisj The gain is small, but the indexer is probably going to be the "hottest path" here.

IMHO the size limit is so old that probably a lot of system parts rely on it and it is not going to change in the future. If it changes, the whole idea of Slice is to have small pieces of huge vector, so the Length should be much smaller than Max, because it makes no sense otherwise. And even if somebody abuses that in the future he or she is going to get some item from the collection instead of Exception so we keep the Memory Safety. 

List<T> also relies on that: http://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,0c65bec4d3fac21e

I think that we can keep it, but if you wish I can revert this change easily.